mountpoint-s3/src/data_cache/mod.rs

```diff
@@ -0,0 +1,54 @@
+//! Traits and types for data caching.
+//!
+//! The data cache aims to reduce repeated fetches of S3 object content,
+//! reducing both the number of requests as well as the latency for the reads.
+//! Ultimately, this means reduced cost in terms of S3 billing as well as compute time.
+
+mod in_memory_data_cache;
+
+use std::ops::Range;
+
+use thiserror::Error;
+
+pub use crate::prefetch::checksummed_bytes::ChecksummedBytes;
+
+/// Indexes blocks within a given object.
+pub type BlockIndex = u64;
+
+/// Errors returned by operations on a [DataCache]
+#[derive(Debug, Error)]
+pub enum DataCacheError {
+    /// It was not possible to read from the cache
+    #[error("Failed reading or writing from cache: {0}")]
+    IoFailure(#[source] anyhow::Error),
+}
+
+pub type DataCacheResult<Value> = Result<Value, DataCacheError>;
+
+/// Cache data with a checksum identified by some [Key].
+///
+/// The underlying cache is divided into blocks of equal size.
+///
+/// TODO: Deletion and eviction of cache entries.
+/// TODO: Some version information (ETag) independent from [Key] to allow smarter eviction?
+pub trait DataCache<Key> {
+    /// Get block of data from the cache for the given [Key] and [BlockIndex], if available.
+    ///
+    /// Operation may fail due to errors, or return [None] if the block was not available in the cache.
+    fn get_block(&self, cache_key: &Key, block_idx: BlockIndex) -> DataCacheResult<Option<&ChecksummedBytes>>;
+
+    /// Put block of data to the cache for the given [Key] and [BlockIndex].
+    fn put_block(&mut self, cache_key: Key, block_idx: BlockIndex, bytes: ChecksummedBytes) -> DataCacheResult<()>;
```

Comment:
Should we require interior mutability for that data cache? I don't think we want the caller to lock the whole cache, even more for a future disk cache.

Reply:
Yes! Big oversight, let me fix :) I'll update the in-mem test impl to just have a `RwLock<>` over the whole cache, but we can do much better with the on disk implementation.

Reply:
We can provide range/extent lock for parallel write。o

Reply:
Thanks! For now since it is for testing only, I don't plan to add more complexity to the in-memory implementation of `DataCache` trait but we will certainly be exploring how we can lock only a subset of the on-disk cache, such as individual blocks.